storage/raft_log_queue.go

```diff
@@ -0,0 +1,127 @@
+	raftLogQueueMaxSize = 100
+	// RaftLogQueueTimerDuration is the duration between checking the
+	// raft logs.
+	RaftLogQueueTimerDuration = time.Second
+	// RaftLogQueueLogSizeThreshold is the number of raft log entries after
+	// which truncation is required.
+	RaftLogQueueLogSizeThreshold = 1
+)
+
+// raftLogQueue manages a queue of replicas slated to have their raft logs
+// truncated by removing unneeded entries.
+type raftLogQueue struct {
+	baseQueue
+	db *client.DB
+}
+
+// newRaftLogQueue returns a new instance of raftLogQueue.
+func newRaftLogQueue(db *client.DB, gossip *gossip.Gossip) *raftLogQueue {
+	rlq := &raftLogQueue{
+		db: db,
+	}
+	rlq.baseQueue = makeBaseQueue("raftlog", rlq, gossip, raftLogQueueMaxSize)
+	return rlq
+}
+
+func (*raftLogQueue) needsLeaderLease() bool {
+	return true
+}
+
+func (*raftLogQueue) acceptsUnsplitRanges() bool {
+	return true
+}
+
+// getTruncatableIndexes looks up the replica's raft status and returns the
+// current first index for the replica and the oldest still in use index.
+func getTruncatableIndexes(r *Replica) (uint64, uint64, error) {
+	desc := r.Desc()
+	raftStatus := r.store.RaftStatus(desc.RangeID)
+	if raftStatus == nil {
+		return 0, 0, util.Errorf("could not get raft status for %s", desc)
+	}
+	firstIndex, err := r.FirstIndex()
+	if err != nil {
+		return 0, 0, util.Errorf("error retrieving first index for range %s: %s", desc, err)
+	}
+
+	// Find the oldest index still in use by the range.
+	oldestIndex := raftStatus.Applied
+	for _, progress := range raftStatus.Progress {
+		if progress.Match < oldestIndex {
+			oldestIndex = progress.Match
+		}
+	}
+
+	return firstIndex, oldestIndex, nil
+}
+
+// shouldQueue determines whether a range should be queued for truncating. This
+// is true if the range's raft log's old entries exceeds the size threshold.
+func (*raftLogQueue) shouldQueue(now roachpb.Timestamp, r *Replica, _ *config.SystemConfig) (shouldQ bool,
+	priority float64) {
+	firstIndex, oldestIndex, err := getTruncatableIndexes(r)
+	if err != nil {
+		log.Warning(err)
+		return false, 0
+	}
+	// Can and should the raft logs be truncated?
+	truncatableIndexes := oldestIndex - firstIndex
+	return truncatableIndexes > RaftLogQueueLogSizeThreshold, float64(truncatableIndexes)
+}
+
+// process synchronously invokes for each replica that requires a raft log
+// truncation.
+func (*raftLogQueue) process(now roachpb.Timestamp, r *Replica, _ *config.SystemConfig) error {
+	firstIndex, oldestIndex, err := getTruncatableIndexes(r)
+	if err != nil {
+		return err
+	}
+
+	// Can and should the raft logs be truncated?
+	truncatableIndexes := oldestIndex - firstIndex
+	if truncatableIndexes > RaftLogQueueLogSizeThreshold {
+		if log.V(1) {
+			log.Infof("truncating the raft log of range %d from %d to %d", r.Desc().RangeID, firstIndex, oldestIndex)
+		}
+		_, err := client.SendWrapped(r, r.context(), &roachpb.TruncateLogRequest{Index: oldestIndex})
```

Comment:
This needs to be sent through `rlq.db` instead of `r`, so the client layer can handle any `NotLeaderErrors` that are returned (because here we need the lease, not raft leadership)